mcstatus/forge_data.py

```diff
@@ -0,0 +1,183 @@
+"""Forge Data Decoder
+
+Forge mod data is encoded into a UTF-16 string that represents
+binary data containing data like the forge mod loader network
+version, a big list of channels that all the forge mods use,
+and a list of mods the server has.
+
+For more information see this file from forge itself:
+https://github.com/MinecraftForge/MinecraftForge/blob/42115d37d6a46856e3dc914b54a1ce6d33b9872a/src/main/java/net/minecraftforge/network/ServerStatusPing.java
+"""
+
+from __future__ import annotations
+
+import io
+from dataclasses import dataclass
+from typing import Final, TYPE_CHECKING
+
+from mcstatus.protocol.connection import Connection
+
+VERSION_FLAG_IGNORE_SERVER_ONLY: Final = 0b1
+IGNORE_SERVER_ONLY: Final = "<not required for client>"
+
+
+if TYPE_CHECKING:
+    from typing_extensions import NotRequired, Self, TypedDict
+
+    class ForgeDataChannel(TypedDict):
+        res: str
+        """Channel name and ID (for example ``fml:handshake``)."""
+        version: str
+        """Channel version (for example ``1.2.3.4``)."""
+        required: bool
+        """Is this channel required for client to join?"""
+
+    class ForgeDataMod(TypedDict):
+        modid: str
+        modmarker: str
+        """Mod version"""
+
+    class RawForgeData(TypedDict):
+        fmlNetworkVersion: int  # noqa: N815
+        channels: list[ForgeDataChannel]
+        mods: list[ForgeDataMod]
+        d: NotRequired[str]
+        truncated: NotRequired[bool]
+
+else:
+    ForgeDataChannel = dict
+    ForgeDataMod = dict
+    RawForgeData = dict
+
+
+@dataclass
+class ForgeData:
+    fml_network_version: int
+    """Forge Mod Loader network version."""
+    channels: list[ForgeDataChannel]
+    """List of channels, both for mods and non-mods."""
+    mods: list[ForgeDataMod]
+    """List of mods"""
+    truncated: bool
+    """Is the mods list and or channel list incomplete?"""
+
+    @classmethod
+    def build(cls, raw: RawForgeData) -> Self:
+        """Build :class:`ForgeData` from raw response :class:`dict`.
+
+        :param raw: Raw forge data response :class:`dict`.
+        :return: :class:`ForgeData` object.
+        """
+        raw.setdefault("fmlNetworkVersion", 0)
+        raw.setdefault("channels", [])
+        raw.setdefault("mods", [])
```

Comment:
I am unsure, that we should modify incoming dict. Instead, it's better to parse these objects into dataclasses.

Reply:
Taken care of in 0ea7af0